### DISCUSSION TITLE
How to export node position

### DISCUSSION THREAD

**Discussion by linonetwo** [retejs/rete#620] [Q&A] [answered]
`node.x` seem not to exist.

**Comment by Ni55aN** [comment, marked as answer]
```ts
area.nodeViews.get(nodeId).position // { x, y }, for v2
node.position // [x, y], for v1
```

**Reply by linonetwo** [reply]
Thanks, this works.

1. I can't get area from editor (though I set it using `editor.use(area)`).
2. This means area is also a top-level concept with editor? And other plugins are also top-level concept in parallel with editor? I think this is not very clear in the guide (https://retejs.org/docs/concepts/editor#create-2d-area).
1. And when setting position back, I just do `area.nodeViews.get(nodeId).position = parsedJSON.position`?

**Reply by linonetwo** [reply]
After reading the code

1. is impossible, because [`class Scope`](https://github.com/retejs/rete/blob/befaa46c597e89e2df74918168e4ae84643dff14/src/scope.ts#L24) don't have getChildByName or something similar, and is not intended to.
2. It is said in https://retejs.org/docs/concepts/plugin-system, editor and other plugins are all scope (can be regarded as a home made event emitter), and form a tree of scope
3. trying

**Reply by linonetwo** [reply]
@Ni55aN This will say `Type '{ position: { x: number; y: number; }; }' is missing the following properties from type 'NodeView': getZoom, events, guards, element, and 4 more.ts(2345)`

I think set method should accept Partial value.

```ts
function applyNodeInfoToArea(node: ReteNode, nodeData: INodeExportedJSON, area: AreaPlugin<EditorSchemes, AreaExtra>) {
  if (nodeData.x !== undefined && nodeData.y !== undefined) {
    area.nodeViews.set(node.id, { position: { x: nodeData.x, y: nodeData.y } });
  }
}
```

I have to do this as workaround

```ts
function applyNodeInfoToArea(node: ReteNode, nodeData: INodeExportedJSON, area: AreaPlugin<EditorSchemes, AreaExtra>) {
  const previousValue = area.nodeViews.get(node.id);
  if (nodeData.x !== undefined && nodeData.y !== undefined && previousValue !== undefined) {
    previousValue.position = { x: nodeData.x, y: nodeData.y };
    area.nodeViews.set(node.id, previousValue);
  }
}
```

**Reply by Ni55aN** [reply]
You cannot change the position in this way. You need area.translate

**Reply by linonetwo** [reply]
Thanks, finally:

```ts
async function applyNodeInfoToArea(node: ReteNode, nodeData: INodeExportedJSON, area: AreaPlugin<EditorSchemes, AreaExtra>) {
  if (nodeData.x !== undefined && nodeData.y !== undefined) {
    await area.translate(node.id, { x: nodeData.x, y: nodeData.y });
  }
}
```

Do this after adding the node

```ts
// for ...
    const task = editor.addNode(node).then(async () => {
      await applyNodeInfoToArea(node, nodeData, area);
    });
    addNodeTasks.push(task);
  }
  await Promise.all(addNodeTasks);
```

**Reply by linonetwo** [reply]
Sorry, I should have lerant this from https://retejs.org/examples/performance

**Reply by Ni55aN** [reply]
concurrent `editor.addNode` may not work correctly in some cases. It's better to avoid it if in the future you have to use some plugins or make your own features that depend on the order of adding nodes

**Reply by Ni55aN** [reply]
np, I'll add more info in docs about that
